src/Lean/Elab/Tactic/BuiltinTactic.lean

```diff
@@ -35,55 +35,52 @@ open Parser.Tactic
 @[builtinTactic paren] def evalParen : Tactic := fun stx =>
   evalTactic stx[1]
 
-def isCheckpointableTactic (arg : Syntax) : TacticM Bool := do
+def isCheckpointableTactic (arg : Syntax.Tactic) : TacticM Bool := do
   -- TODO: make it parametric
-  let kind := arg.getKind
+  let kind := arg.1.getKind
   return kind == ``Lean.Parser.Tactic.save
 
-partial def addCheckpoints (args : Array Syntax) : TacticM (Array Syntax) := do
+partial def addCheckpoints (args : Array Syntax.Tactic) : TacticM (Array Syntax.Tactic) := do
   -- if (← readThe Term.Context).tacticCache? |>.isSome then
-  if (← args.anyM fun arg => isCheckpointableTactic arg[0]) then
+  if (← args.anyM fun arg => isCheckpointableTactic arg) then
     let argsNew ← go 0 #[] #[]
     return argsNew
   return args
 where
-  push (acc : Array Syntax) (result : Array Syntax) : Array Syntax :=
+  push (acc : Array Syntax.Tactic) (result : Array Syntax.Tactic) : Array Syntax.Tactic :=
     if acc.isEmpty then
       result
     else
-      let ref := acc.back
-      let accSeq := mkNode ``Lean.Parser.Tactic.tacticSeq1Indented #[mkNullNode acc]
-      let checkpoint := mkNode ``Lean.Parser.Tactic.checkpoint #[mkAtomFrom ref "checkpoint", accSeq]
-      result.push (mkNode groupKind #[checkpoint, mkNullNode])
+      result.push <| Unhygienic.run do withRef acc.back do
+        `(tactic| checkpoint $[$acc]*)
 
-  go (i : Nat) (acc : Array Syntax) (result : Array Syntax) : TacticM (Array Syntax) := do
+  go (i : Nat) (acc : Array Syntax.Tactic) (result : Array Syntax.Tactic) : TacticM (Array Syntax.Tactic) := do
     if h : i < args.size then
       let arg := args.get ⟨i, h⟩
-      if (← isCheckpointableTactic arg[0]) then
+      if (← isCheckpointableTactic arg) then
         -- `argNew` is `arg` as singleton sequence. The idea is to make sure it does not satisfy `isCheckpointableTactic` anymore
-        let argNew := arg.setArg 0 (mkNullNode #[arg[0]])
+        let argNew := ⟨mkNullNode #[arg]⟩ -- HACK
         let acc := acc.push argNew
         go (i+1) #[] (push acc result)
       else
         go (i+1) (acc.push arg) result
     else
       return result ++ acc
 
-/-- Evaluate `many (group (tactic >> optional ";")) -/
-def evalManyTacticOptSemi (stx : Syntax) : TacticM Unit := do
-  for seqElem in (← addCheckpoints stx.getArgs) do
```

Comment:
You are converting here to `Array Syntax.Tactic`. It is not clear to me why.

Reply:
(Note: I have rewritten this function one more time since your comment.)

The general reason is that the AST of tactic sequences has changed from `many1 (group (tactic >> optional ";"))` to `sepBy1Indent tactic`, i.e. from `[(group (skip) [])]` to `[(skip) []]`.  Having explicit types available (`Array Syntax.Tactic` instead of `Syntax`) makes the refactoring easier because the type system tells you what is a list of tactics, what is a tactic, etc., and you don't have to guess yourself what the `Syntax` means.

At first it seemed like it was not necessary to keep the separators, making `Array Syntax.Tactic` a natural choice.

I have now changed it back to untyped `Syntax` because we need to preserve the separators.  Ideally we'd have a `TSepIndentArray` for this because the manual index calculations are not particularly pretty.  (However that doesn't solve all problems either.  Even the `TSepArray` API is pretty bare-bones---and it doesn't preserve the separators either.)

Reply:
> `TSepIndentArray`

Oh I completely forgot we had this already.  It's called ``TSepArray `tactic ""``.